File: src/Block.js
const crypto = require("crypto-js");

class Block {
    constructor(index, timestamp, previousHash, data, validator) {
        this.index = index;
        this.timestamp = timestamp;
        this.previousHash = previousHash;
        this.data = data;
        this.validator = validator;
        this.hash = this.calculateHash();
    }

    calculateHash = () => {
        return crypto.SHA256(
            this.index + this.timestamp + this.previousHash + JSON.stringify(this.data) + this.validator
        ).toString();
    }
}

module.exports = Block;

File: lib/src/block.dart
import 'dart:convert';
import 'package:crypto/crypto.dart';

import 'package:netwealth/src/transaction.dart';

class Block {
  int index;
  DateTime timestamp;
  String previousHash;
  List<Transaction> data;
  String validator;
  String hash;

  Block(
      this.index, this.timestamp, this.previousHash, this.data, this.validator)
      : hash = calculateHash(index, timestamp, previousHash, data, validator);

  static String calculateHash(int index, DateTime timestamp,
      String previousHash, List<Transaction> data, String validator) {
    final dataString =
        jsonEncode(data.map((transaction) => transaction.toJson()).toList());
    final input = '$index$timestamp$previousHash$dataString$validator';
    return sha256.convert(utf8.encode(input)).toString();
  }
}
